[ios][background-task] Added showing an error when bg tasks are not available

chrfalch opened the pull request:
# Why

If background tasks aren't available we need to show a message to the user trying to explain whats going on.

Closes #ENG-15092

Usually it will be because the user is trying to run on an iOS simulator or web.

# How

This commit fixes this by throwing an error if you try to call `registerTask` on an iOS device or an unsupported device.

# Test Plan

Run in BareExpo on iOS Simulator and add a call to `registerTask` in `BackgroundTaskScreen.tsx`
Observe that an exception is showed.

# Checklist

- [X] I added a `changelog.md` entry and rebuilt the package sources according to [this short guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [X] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

chrfalch committed:
[ios][background-task] Added showing an error when bg tasks are not available

Closes #ENG-15092

If background tasks aren't available we need to show a message to the user trying to explain whats going on.

Usually it will be because the user is trying to run on an iOS simulator or web.

This commit fixes this by both displaying a message over console.error to ensure that it is visible even though we're in an async method and then also throwing an error to stop further execution.

chrfalch committed:
Update packages/expo-background-task/CHANGELOG.md

Co-authored-by: Expo Bot <34669131+expo-bot@users.noreply.github.com>

chrfalch committed:
removed console.error

brentvatne commented on packages/expo-background-task/src/BackgroundTask.ts:
```diff
@@ -60,6 +60,13 @@ export async function registerTaskAsync(
       `Task '${taskName}' is not defined. You must define a task using TaskManager.defineTask before registering.`
     );
   }
+  if ((await ExpoBackgroundTaskModule.getStatusAsync()) === BackgroundTaskStatus.Restricted) {
```
maybe we should warn instead of error, and just noop

chrfalch replied:
Good point!

chrfalch committed:
codereview: implemented a warning instead of error

Now a warning is displayed if you try to call the method in a non-supported environment. A specific message is displayed for iOS.

Updated setting registered after starting/stopping a task

chrfalch committed:
Refactored variable name

chrfalch committed:
rephrasing

brentvatne commented on packages/expo-background-task/src/BackgroundTask.ts:
```diff
@@ -60,6 +63,17 @@ export async function registerTaskAsync(
+    if (!warnAboutRunningOniOSSimulator) {
+      const message =
+        Platform.OS === 'ios'
+          ? 'On iOS, expo-background-tasks can only be used on a physical device and is not supported on the simulator.'
```
```suggestion
          ? `Background tasks are not supported on iOS simulators. Skipped registering task: ${taskName}.`
```

how about something like this? if you like it, we could update the other message ("Background tasks are not available in the current environment") to add the task name as well

chrfalch replied:
I like it - which other message are you thinking about?

brentvatne replied:
looks like you got the other msg, i meant the one right right below the first one haha. sorry i wasn't clear

chrfalch committed:
Update packages/expo-background-task/src/BackgroundTask.ts

Co-authored-by: Brent Vatne <brentvatne@gmail.com>

chrfalch committed:
codereview: Built and update other message
